File: code/Python/Chapter4/4-3/postfix.py
def calculate(expression):
    stack = []
    tokens = expression.split()

    for token in tokens:
        if token in "+-*/":
            right = stack.pop()
            left = stack.pop()
            if token == "+":
                stack.append(left + right)
            elif token == "-":
                stack.append(left - right)
            elif token == "*":
                stack.append(left * right)
            elif token == "/":
                stack.append(left / right)
        else:
            stack.append(int(token))
    
    return stack.pop()

expressions = [
    "1 2 +",            # 1 + 2 = 3
    "2 3 4 + *",        # 2 * (3 + 4) = 14
    "1 2 + 3 4 + *",    # (1 + 2) * (3 + 4) = 21
    "3 4 2 + * 5 *",    # 3 * (4 + 2) * 5 = 90
    "50 20 - 2 /",      # (50 - 20) / 2 = 15
]

for expression in expressions:
    print(expression, "=", calculate(expression))

File: code/Java/Chapter4/4-3/Postfix.java
import java.util.Stack;

public class Postfix {
    public static void main(String[] args) {
        String[] expressions = {
                "1 2 +",            // 1 + 2 = 3
                "2 3 4 + *",        // 2 * (3 + 4) = 14
                "1 2 + 3 4 + *",    // (1 + 2) * (3 + 4) = 21
                "3 4 2 + * 5 *",    // 3 * (4 + 2) * 5 = 90
                "50 20 - 2 /",      // (50 - 20) / 2 = 15
        };

        for (String expression : expressions) {
            System.out.println(expression + " = " + calculate(expression));
        }
    }

    public static double calculate(String expression) {
        Stack<Double> stack = new Stack<Double>();
        String[] tokens = expression.split(" ");

        for (String token : tokens) {
            if (token.equals("+")) {
                double right = stack.pop();
                double left = stack.pop();
                stack.push(left + right);
            } else if (token.equals("-")) {
                double right = stack.pop();
                double left = stack.pop();
                stack.push(left - right);
            } else if (token.equals("*")) {
                double right = stack.pop();
                double left = stack.pop();
                stack.push(left * right);
            } else if (token.equals("/")) {
                double right = stack.pop();
                double left = stack.pop();
                stack.push(left / right);
            } else {
                stack.push(Double.parseDouble(token));
            }
        }

        return stack.pop();
    }
}